scrapinghub/client.py

```diff
@@ -0,0 +1,370 @@
+        return self._client._connection.project_ids()
+
+    @wrap_http_errors
+    def summary(self, **params):
+        return self._client._hsclient.projects.jobsummaries(**params)
+
+
+class Project(object):
+
+    def __init__(self, client, projectid):
+        self.id = projectid
+        self._client = client
+
+        # sub-resources
+        self.jobs = Jobs(client, projectid)
+        self.spiders = Spiders(client, projectid)
+
+        # proxied sub-resources
+        self.activity = Activity(_Activity, client, projectid)
+        self.collections = Collections(_Collections, client, projectid)
+        self.frontier = Frontier(client._hsclient, projectid)
+        self.settings = Settings(client._hsclient, projectid)
+
+
+class Spiders(object):
+
+    def __init__(self, client, projectid):
+        self.projectid = projectid
+        self._client = client
+
+    @wrap_http_errors
+    def get(self, spidername, **params):
+        project = self._client._hsclient.get_project(self.projectid)
+        spiderid = project.ids.spider(spidername, **params)
+        if spiderid is None:
+            raise NonExistingSpider("Spider {} doesn't exist."
+                                    .format(spidername))
+        return Spider(self._client, self.projectid, spiderid, spidername)
+
+    def list(self):
+        project = self._client._connection[self.projectid]
+        return project.spiders()
+
+
+class Spider(object):
+
+    def __init__(self, client, projectid, spiderid, spidername):
+        self.projectid = projectid
+        self.id = spiderid
+        self.name = spidername
+        self.jobs = Jobs(client, projectid, self)
+        self._client = client
+
+    def update_tags(self, add=None, remove=None):
+        params = get_tags_for_update(add_tag=add, remove_tag=remove)
+        if not params:
+            return
+        params.update({'project': self.projectid, 'spider': self.name})
+        result = self._client._connection._post('jobs_update', 'json', params)
+        return result['count']
+
+
+class Jobs(object):
+
+    def __init__(self, client, projectid, spider=None):
+        self.projectid = projectid
+        self.spider = spider
+        self._client = client
+        self._project = client._hsclient.get_project(projectid)
+
+    @wrap_http_errors
+    def count(self, **params):
+        if self.spider:
+            params['spider'] = self.spider.name
+        return next(self._project.jobq.apiget(('count',), params=params))
+
+    @wrap_http_errors
+    def iter(self, **params):
+        """ Iterate over jobs collection for a given set of params.
+        FIXME the function returns a list of dicts, not a list of Job's
+        """
+        if self.spider:
+            params['spider'] = self.spider.name
+        return self._project.jobq.list(**params)
+
+    def schedule(self, spidername=None, **params):
+        if not spidername and not self.spider:
+            raise APIError('Please provide spidername')
```

Comment:
I'd say it should be a `ValueError`, `APIError` now mixes errors from Connection (HTTP errors?), errors from new client. IMO:

- `ValueError` is what needed if method receives unexpected argument (or doesn't receive expected)
- new client deserves new base Exception class, new exceptions like DuplicateJobError should inherit from new base class - let's keep APIError for legacy Connection 
- Most of 4xx responses should be covered with specific exceptions classes, like 400 should result in `ValidationError` or `InvalidUsage`, 404 in generic `NotFound`. they also should store either original `requests.exceptions.HTTPError` or response object - to be able to get body and check why things failed.
- 5xx errors should be raised as is I think, we just need to be sure it's always requests.exceptions.HTTPError